mq/plugins/broFixup.py

```diff
@@ -41,39 +45,493 @@ def __init__(self):
+            pass
+
 
     def onMessage(self, message, metadata):
         # set the doc type to bro
         # to avoid data type conflicts with other doc types
         # (int v string, etc)
-        metadata['doc_type']= 'bro'
+        # index holds documents of type 'type'
+        # index -> type -> doc
+        metadata['doc_type']= 'nsm'
+        whitelist = ['hostname', 'type', 'tags', 'customendpoint']
+
+        # move Bro specific fields under 'details' while preserving metadata
+        newmessage = dict()
+
+        newmessage['details'] = message
+        newmessage['customendpoint'] = 'bro'
+
+
+        # add mandatory fields
+        if 'ts' in message.keys():
+            newmessage[u'utctimestamp'] = toUTC(message['ts']).isoformat()
+            newmessage[u'timestamp'] = toUTC(message['ts']).isoformat()
+        else:
+            # a malformed message somehow managed to crawl to us, let's put it somewhat together
+            newmessage[u'utctimestamp'] = toUTC(datetime.now()).isoformat()
+            newmessage[u'timestamp'] = toUTC(datetime.now()).isoformat()
+
+        newmessage[u'receivedtimestamp'] = toUTC(datetime.now()).isoformat()
+
+
+        # add mandatory fields
+        # move bro data under details
+        newmessage[u'eventsource'] = u'nsm'
+        newmessage[u'severity'] = u'INFO'
+        #newmessage[u'processname'] = u''
+        #newmessage[u'processid'] = u''
+        newmessage[u'mozdefhostname'] = self.mozdefhostname
+        newmessage[u'category'] = u'bro'
+
 
         # re-arrange the position of some fields
         # {} vs {'details':{}}
-        if 'details' in message.keys():
-            # details.tags -> tags
-            if 'tags' in message['details'].keys():
-                message['tags'] = message['details']['tags']
-                del message['details']['tags']
-
-            # details.summary -> summary
-            if 'summary' in message['details'].keys():
-                message['summary'] = message['details']['summary']
-                del message['details']['summary']
-
-            # clean up the action notice IP addresses
-            if 'actions' in message['details'].keys():
-                if message['details']['actions'] == "Notice::ACTION_LOG":
-                    # retrieve indicator ip addresses from the sub field
-                    # "sub": "Indicator: 1.2.3.4, Indicator: 5.6.7.8"
-                    message['details']['indicators'] = [ip for ip
-                                                        in findIPv4(message['details']['sub'])]
+        if 'details' in newmessage.keys():
+
+            #newmessage[u'hostname'] = newmessage['details']['hostname']
+            #del(newmessage['details']['hostname'])
+
+            # some Bro logs need special treatment
+            if 'type' in newmessage['details']:
+                logtype = newmessage['details']['type']
+
+                if logtype == 'conn':
```

Comment:
I think unit tests here for each 'logtype' would really help clarify the transformations that we're doing. It will also help lock in the functionality so if we ever need to go in here to modify stuff, we can ensure we aren't unexpectedly breaking anything.